SentenceVectors.ecl

```diff
@@ -0,0 +1,651 @@
+  *                     10 and 100 have an edit distance of only 1 (i.e. add a zero), yet the
+  *                     meaning can be vastly different.  For numeric words, we use the actual
+  *                     numeric distance instead of edit distance.  For example, if this parameter
+  *                     is set to 9, then 43 will match to 50 if it is the closest number.
+  *                     A value of zero disables numeric matching (Default 9).
+  * @param saveSentences Set to TRUE to save the training sentences in the model. It should always
+  *                        be set to TRUE if you plan to call GetClosestSentences(...).  Setting
+  *                        to FALSE will result in a smaller model with only the word vectors
+  *                        (default TRUE).
+  */
+EXPORT SentenceVectors(UNSIGNED2 vecLen=100,
+                      REAL trainToLoss = .05,
+                      UNSIGNED4 numEpochs= 0,
+                      UNSIGNED4 batchSize=0,
+                      UNSIGNED4 negSamples=10,
+                      REAL4 learningRate=.2,
+                      REAL4 discardThreshold = .0001,
+                      UNSIGNED4 minOccurs = 10,
+                      UNSIGNED4 wordNGrams = 1,
+                      UNSIGNED4 dropoutK = 3,
+                      UNSIGNED4 noProgressEpochs = 1,
+                      UNSIGNED4 maxTextDistance = 3,
+                      UNSIGNED4 maxNumDistance = 9,
+                      BOOLEAN saveSentences=TRUE) := MODULE
+  // Calibration constant (heuristic) for adjusting batch size
+  SHARED calConst := 25;
+  /**
+    * Extract the word vectors from the trained weights
+    */
+  SHARED computeWordVectors(DATASET(SliceExt) slices, DATASET(WordInfo) words,
+																SET OF UNSIGNED shape) := FUNCTION
+    w := int.Weights(shape);  // Module to manage weights
+    // As an optimization, we only convert half of the weight slices
+    // since the word vectors are always in the first half
+    firstHalf := ROUNDUP(w.nSlices / 2);
+    allWeights := w.slices2Linear(slices(sliceId <= firstHalf ));
+    // Extract the vectors for each word.  Words should be evenly distributed at this point.
+    WordInfo getVectors(WordInfo wrd) := TRANSFORM
+      // The weights that form the word vector are the layer 1 weights.  j is the wordId
+      // and i is the term number of the word vector.  The terms of the word vector are
+      // contiguous, so we only need to know the start and end indexes.
+      startIndx := w.toFlatIndex(1, wrd.wordId, 1);
+      endIndx := startIndx + shape[2] - 1;
+      SELF.vec := Utils.normalizeVector(allWeights.weights[startIndx .. endIndx]);
+      SELF := wrd;
+    END;
+    wordsWithVecs := PROJECT(words, getVectors(LEFT), LOCAL);
+    RETURN wordsWithVecs;
+  END;
+  // This function is an alternative to computeWordVectors above.  I need to compare the two
+  // to see which one performs better -- TODO*
```

Comment:
Please consider to remove the TODO* in comments.

Reply:
deferred